Provide custom grpc/http client for DaprClientBuilder as in #683 

subash89 opened the pull request:
Signed-off-by: Subash Gamage <subashsdm@gmail.com>

# Description
This is to Enhance DaprClientBuilder to override gRPC and Http Client. Complete description and reference JIRA can be found at https://github.com/dapr/java-sdk/issues/683 

This includes changes in DaprClientBuilder class to accept a managed grpc client as well as ok http client.

## Issue reference

https://github.com/dapr/java-sdk/issues/683

Please reference the issue this PR will close: #_[issue number]_

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation

artursouza commented on README.md:
````diff
@@ -160,6 +160,27 @@ This SDK provides a basic serialization for request/response objects but also fo
+
+    ```java 
+       DaprClientBuilder daprClientBuilder = new DaprClientBuilder();
+       DaprClient client = daprClientBuilder.withOkHttpClient(new OkHttpClient.Builder().readTimeout(5, TimeUnit.SECONDS)).build();
````
We actually plan to remove the dependency on OkHTTP. Adding that to our public methods will make that harder. Why can't a custom gRPC or HTTP client be implemented with the existing interfaces?

subash89 replied:
The DaprClientBuilder create the channel internally and not allowing a way to override that outside.

artursouza commented on README.md:
````diff
@@ -160,6 +160,27 @@ This SDK provides a basic serialization for request/response objects but also fo
+
+    ```java
+        DaprClientBuilder daprClientBuilder = new DaprClientBuilder();
+        ManagedChannelBuilder managedChannelBuilder = Grpc.newChannelBuilder("localhost:8000", TlsChannelCredentials.create())
````
This is interesting. Which are the most common settings for a managed channel? Can this be done via additional java properties?

subash89 replied:
Channel has everything that matters and as I know we cannot control them via out of the box sys properties. Example using my own managed executor service

subash89 commented:
@artursouza  I updated the PR with new changes and reverted the old commit. New change is minimal as  I provided a easy way to extend the builder class. Because maintaining a two Builder classes is hard to maintain as code is duplicated. 
Let me know your thoughts. And I did not touch http builder yet as you said okhttp will go away. Once we have that finalized, I will add capability to override http client as well.

mukundansundar reviewed (commented):
@subash89 Please sign off your commits following : https://github.com/dapr/java-sdk/blob/master/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work

subash89 commented:
Looks like it didn't like my reverse commit. I followed 
git checkout <branch-name>
git commit --amend --no-edit --signoff
git push --force-with-lease <remote-name> <branch-name>
@mukundansundar  Is it looks good now ? (How to make sure DCO is all set on a given pull request ?)

mukundansundar commented:
@subash89 Can you rebase all your commits into a single commit ? Also could you fix the check style violations in the PR?

subash89 commented:
@mukundansundar 
If you mean to say squash commits, I cannot squash them into one as I already pushed them to the branch. If not can you suggest how can I do that ? Or I can do a new PR with only this change. 

This is my first PR, where to find checkstyle report inorder to fix those.

artursouza commented on sdk/src/main/java/io/dapr/client/DaprExtensibleClientBuilder.java:
```diff
@@ -0,0 +1,29 @@
+/**
+ * A builder for the DaprClient.
+ */
+public class DaprExtensibleClientBuilder extends DaprClientBuilder{
```
I would make this a GRPC specific builder and force the GRPC protocol, even if property is for HTTP.

With the current design, it would be possible for the code to set the GRPC managed channel and a Java property force the client built to be the HTTP instance.

subash89 replied:
Make sense.

subash89 commented:
@artursouza  meantime I update the PR, can you pls suggest a way to build this on a mac or some easy way to find out what are the checkstyle violaitons are.

subash89 commented:
@artursouza Updated the PR. please let me know.

subash89 commented:
@artursouza  any help appreciated.

mukundansundar commented:
> @artursouza any help appreciated.

@subash89 Can you merge with latest master resolve conflicts and see if you are able to build it with the changes in master?

subash89 commented:
@mukundansundar main problem I am having is to build this in a mac?
[ERROR] Failed to execute goal com.github.os72:protoc-jar-maven-plugin:3.11.4:run (default) on project dapr-sdk-autogen: Error extracting protoc for version 3.13.0: Unsupported platform: protoc-3.13.0-osx-x86_64.exe -> [Help 1]

mukundansundar commented:
> @mukundansundar main problem I am having is to build this in a mac?
> [ERROR] Failed to execute goal com.github.os72:protoc-jar-maven-plugin:3.11.4:run (default) on project dapr-sdk-autogen: Error extracting protoc for version 3.13.0: Unsupported platform: protoc-3.13.0-osx-x86_64.exe -> [Help 1]

Does this happen with the latest changes from master too?

subash89 commented:
Yes in master too

subash89 commented:
Can this be build on a linux box ?

mukundansundar commented:
> Can this be build on a linux box ?

the build workflow is on a linux runner only

amuthapriya4 commented:
Please commit the changes for Managed channel to the release branch. With current version it going with default config.

cicoyle commented:
@subash89 - mind resolving the conflicts?

cicoyle commented:
gentle ping @subash89

artur-ciocanu commented:
@artursouza and @subash89 is this PR relevant, we don't use OkHttpClient anymore. Could we close it?

salaboy commented:
closing due inactivity @artur-ciocanu @cicoyle
